core/src/main/scala/com/gu/tableversions/core/Version.scala

```diff
@@ -0,0 +1,60 @@
+package com.gu.tableversions.core
+
+import java.time.{Instant, LocalDateTime, ZoneId}
+import java.time.format.DateTimeFormatter
+import java.util.UUID
+
+import cats.effect.IO
+import cats.syntax.either._
+
+import scala.util.Try
+
+/**
+  * Type that represents a valid version, as used by tables and partitions.
+  */
+final case class Version(timestamp: Instant, uuid: UUID) {
+
+  import Version._
+
+  def label: String = {
+    val versionString = uuid
+    val localDateTime = LocalDateTime.ofInstant(timestamp, ZoneId.of("UTC"))
+    val timestampStr = timestampFormatter.format(localDateTime)
+    s"$timestampStr-$versionString"
+  }
+
+}
+
+object Version {
+
+  val Unversioned = Version(Instant.MIN, new UUID(0L, 0L))
+
+  /** Generator for versions using a timestamp + UUID format */
+  implicit val generateVersion: IO[Version] =
+    IO { Version(Instant.now(), UUID.randomUUID()) }
+
+  /** Regex that can be used to recognise a valid format string. */
+  val TimestampAndUuidRegex = """(\d{8}-\d{6}.\d{9})-(.*)""".r
```

Comment:
we have added nanoseconds 👍

Reply:
Yes... for non-obvious reasons...

We only had seconds at first, but then we realised we need to reconstruct a `Version` object from the string label. So the choice was either including the nanoseconds in the string label, or make sure that the `Version` object didn't have nanosecond values in their timestamps - otherwise the roundtrip of object -> string label -> object didn't result in the original object!

So we did the former, still not sure if that's ideal but...